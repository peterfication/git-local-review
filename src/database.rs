use refinery::config::{Config, ConfigDbType};
use sqlx::{SqlitePool, sqlite::SqliteConnectOptions};

use crate::migrations;

pub struct Database {
    pool: SqlitePool,
}

impl Database {
    pub async fn new() -> color_eyre::Result<Self> {
        std::fs::create_dir_all("tmp")?;

        let options = SqliteConnectOptions::new()
            .filename("tmp/reviews.db")
            .create_if_missing(true);

        let pool = SqlitePool::connect_with(options).await?;

        // Run migrations using refinery
        let mut config = Config::new(ConfigDbType::Sqlite).set_db_path("tmp/reviews.db");
        migrations::runner()
            .run(&mut config)
            .map_err(|e| color_eyre::eyre::eyre!("Migration failed: {}", e))?;

        log::info!("Database initialized at tmp/reviews.db with migrations");

        Ok(Self { pool })
    }

    #[cfg(test)]
    pub fn from_pool(pool: SqlitePool) -> Self {
        Self { pool }
    }

    pub fn pool(&self) -> &SqlitePool {
        &self.pool
    }

    pub async fn close(self) -> color_eyre::Result<()> {
        self.pool.close().await;
        Ok(())
    }
}
