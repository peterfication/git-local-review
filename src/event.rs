use std::sync::Arc;
use std::time::Duration;

use color_eyre::eyre::OptionExt;
use futures::{FutureExt, StreamExt};
use ratatui::crossterm::event::{Event as CrosstermEvent, KeyEvent};
use tokio::sync::mpsc;

use crate::{
    models::Review,
    services::state_service::{AppState, StateEvent},
    services::{GitBranchesLoadingState, ReviewCreateData, ReviewsLoadingState},
    views::KeyBinding,
};

/// Type alias for review identifiers to make event signatures more descriptive.
pub type ReviewId = str;

/// The frequency at which tick events are emitted.
const TICK_FPS: f64 = 30.0;

/// Representation of all possible events.
#[derive(Clone, Debug)]
pub enum Event {
    /// An event that is emitted on a regular schedule.
    ///
    /// Use this event to run any code which has to run outside of being a direct response to a user
    /// event. e.g. polling exernal systems, updating animations, or rendering the UI based on a
    /// fixed frame rate.
    Tick,
    /// Crossterm events.
    ///
    /// These events are emitted by the terminal.
    Crossterm(CrosstermEvent),
    /// Application events.
    ///
    /// Use this event to emit custom events that are specific to your application.
    App(AppEvent),
}

/// Application events.
///
/// You can extend this enum with your own custom events.
#[derive(Clone, Debug)]
pub enum AppEvent {
    /// Initialization event.
    Init,
    /// Quit the application.
    Quit,
    /// Close the current view.
    ViewClose,

    /// Trigger loading of reviews.
    ReviewsLoad,
    /// Load the reviews from the database
    ReviewsLoading,
    /// Propagates the current loading state of reviews.
    ReviewsLoadingState(ReviewsLoadingState),
    /// Load a single review by ID.
    ReviewLoad(Arc<ReviewId>),
    /// Single review loaded successfully.
    ReviewLoaded(Arc<Review>),
    /// Review was not found.
    ReviewNotFound(Arc<ReviewId>),
    /// Error occurred while loading a review.
    ReviewLoadError(Arc<ReviewId>),
    /// Inform that a review has been created.
    ReviewCreated(Review),
    /// Error occurred while creating a review.
    ReviewCreatedError(Arc<str>),
    /// Delete the selected review.
    ReviewDelete(Arc<ReviewId>),
    /// Inform that a review has been deleted.
    ReviewDeleted,
    /// Error occurred while deleting a review.
    ReviewDeletedError(Arc<ReviewId>),

    /// Open help modal with keybindings.
    HelpOpen(Arc<[KeyBinding]>),
    /// Key selected from help modal.
    HelpKeySelected(Arc<KeyEvent>),

    /// Open the review creation view.
    ReviewCreateOpen,
    /// Submit the review creation form.
    ReviewCreateSubmit(Arc<ReviewCreateData>),

    /// Open delete confirmation dialog for selected review.
    ReviewDeleteConfirm(Arc<ReviewId>),

    /// Open review details view.
    ReviewDetailsOpen(Arc<ReviewId>),

    /// Trigger loading of Git branches.
    GitBranchesLoad,
    /// Load the Git branches
    GitBranchesLoading,
    /// Propagates the current loading state of Git branches.
    GitBranchesLoadingState(GitBranchesLoadingState),

    /// State service events for centralized state management.
    StateEvent(StateEvent),
    /// Broadcast updated application state to all views.
    StateUpdate(Arc<AppState>),
}

/// Terminal event handler.
#[derive(Debug)]
pub struct EventHandler {
    /// Event sender channel.
    sender: mpsc::UnboundedSender<Arc<Event>>,
    /// Event receiver channel.
    receiver: mpsc::UnboundedReceiver<Arc<Event>>,
}

impl Default for EventHandler {
    fn default() -> Self {
        panic!("Use EventHandler::new() instead of Default");
    }
}

impl EventHandler {
    /// Constructs a new instance of [`EventHandler`] and spawns a new thread to handle events.
    pub fn new() -> Self {
        let (sender, receiver) = mpsc::unbounded_channel();
        let actor = EventTask::new(sender.clone());
        tokio::spawn(async { actor.run().await });
        Self { sender, receiver }
    }

    /// Constructs a new instance of [`EventHandler`] for testing without spawning the event task.
    /// This allows tests to control event flow manually.
    #[cfg(test)]
    pub fn new_for_test() -> Self {
        let (sender, receiver) = mpsc::unbounded_channel();
        Self { sender, receiver }
    }

    /// Receives an event from the sender.
    ///
    /// This function blocks until an event is received.
    ///
    /// # Errors
    ///
    /// This function returns an error if the sender channel is disconnected. This can happen if an
    /// error occurs in the event thread. In practice, this should not happen unless there is a
    /// problem with the underlying terminal.
    pub async fn next(&mut self) -> color_eyre::Result<Arc<Event>> {
        self.receiver
            .recv()
            .await
            .ok_or_eyre("Failed to receive event")
    }

    /// Queue an app event to be sent to the event receiver.
    ///
    /// This is useful for sending events to the event handler which will be processed by the next
    /// iteration of the application's event loop.
    pub fn send(&mut self, app_event: AppEvent) {
        // Ignore the result as the reciever cannot be dropped while this struct still has a
        // reference to it
        let _ = self.sender.send(Event::App(app_event).into());
    }

    /// Queue an app event to be sent to the event receiver (async version).
    ///
    /// This is useful for sending events from async contexts where you need proper error handling.
    pub async fn send_async(&self, app_event: AppEvent) -> color_eyre::Result<()> {
        self.sender
            .send(Event::App(app_event).into())
            .map_err(|_| color_eyre::eyre::eyre!("Failed to send event"))?;
        Ok(())
    }

    /// Queue a key event to be sent to the event receiver as a crossterm event.
    ///
    /// This is useful for programmatically sending key events that will be processed
    /// through the normal key event handling flow.
    pub fn send_key_event(&mut self, key_event: KeyEvent) {
        let crossterm_event = CrosstermEvent::Key(key_event);
        let _ = self.sender.send(Event::Crossterm(crossterm_event).into());
    }

    /// Check if there are any pending events in the queue.
    /// This is useful for testing to verify that events have been sent.
    #[cfg(test)]
    pub fn has_pending_events(&self) -> bool {
        !self.receiver.is_empty()
    }

    /// Try to receive an event without blocking.
    /// Returns None if no events are available.
    /// This is useful for testing to check what events have been sent.
    #[cfg(test)]
    pub fn try_recv(&mut self) -> Option<Arc<Event>> {
        self.receiver.try_recv().ok()
    }
}

/// A thread that handles reading crossterm events and emitting tick events on a regular schedule.
struct EventTask {
    /// Event sender channel.
    sender: mpsc::UnboundedSender<Arc<Event>>,
}

impl EventTask {
    /// Constructs a new instance of [`Event`].
    fn new(sender: mpsc::UnboundedSender<Arc<Event>>) -> Self {
        Self { sender }
    }

    /// Runs the event thread.
    ///
    /// This function emits tick events at a fixed rate and polls for crossterm events in between.
    async fn run(self) -> color_eyre::Result<()> {
        let tick_rate = Duration::from_secs_f64(1.0 / TICK_FPS);
        let mut reader = crossterm::event::EventStream::new();
        let mut tick = tokio::time::interval(tick_rate);
        loop {
            let tick_delay = tick.tick();
            let crossterm_event = reader.next().fuse();
            tokio::select! {
              _ = self.sender.closed() => {
                break;
              }
              _ = tick_delay => {
                self.send(Event::Tick.into());
              }
              Some(Ok(event)) = crossterm_event => {
                self.send(Event::Crossterm(event).into());
              }
            };
        }
        Ok(())
    }

    /// Sends an event to the receiver.
    fn send(&self, event: Arc<Event>) {
        // Ignores the result because shutting down the app drops the receiver, which causes the send
        // operation to fail. This is expected behavior and should not panic.
        let _ = self.sender.send(event);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use ratatui::crossterm::event::{KeyCode, KeyEventKind, KeyEventState, KeyModifiers};

    #[test]
    fn test_send_key_event() {
        let mut event_handler = EventHandler::new_for_test();

        let key_event = KeyEvent {
            code: KeyCode::Char('q'),
            modifiers: KeyModifiers::empty(),
            kind: KeyEventKind::Press,
            state: KeyEventState::empty(),
        };

        // Send a key event
        event_handler.send_key_event(key_event);

        // Verify the event was sent as a crossterm event
        assert!(event_handler.has_pending_events());
        let event = event_handler.try_recv().unwrap();

        match *event {
            Event::Crossterm(CrosstermEvent::Key(received_key_event)) => {
                assert_eq!(received_key_event.code, KeyCode::Char('q'));
                assert_eq!(received_key_event.modifiers, KeyModifiers::empty());
                assert_eq!(received_key_event.kind, KeyEventKind::Press);
                assert_eq!(received_key_event.state, KeyEventState::empty());
            }
            _ => panic!("Expected crossterm key event, got: {event:?}"),
        }
    }
}
