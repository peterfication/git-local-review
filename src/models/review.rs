use chrono::{DateTime, Utc};
use sqlx::{FromRow, SqlitePool};
use uuid::Uuid;

use crate::time_provider::{SystemTimeProvider, TimeProvider};

#[derive(Debug, Clone, FromRow)]
pub struct Review {
    pub id: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub base_branch: String,
    pub target_branch: String,
}

impl PartialEq for Review {
    fn eq(&self, other: &Self) -> bool {
        self.id == other.id
    }
}

impl Review {
    pub fn new(base_branch: String, target_branch: String) -> Self {
        Self::new_with_time_provider(base_branch, target_branch, &SystemTimeProvider)
    }

    pub fn new_with_time_provider(
        base_branch: String,
        target_branch: String,
        time_provider: &dyn TimeProvider,
    ) -> Self {
        let now = time_provider.now();
        Self {
            id: Uuid::new_v4().to_string(),
            created_at: now,
            updated_at: now,
            base_branch,
            target_branch,
        }
    }

    /// Returns a human-readable title for the review in the format "base_branch -> target_branch"
    pub fn title(&self) -> String {
        format!("{} -> {}", self.base_branch, self.target_branch)
    }

    #[cfg(test)]
    pub fn test_review(opts: impl Into<TestReviewParams>) -> Self {
        let opts = opts.into();
        Self::new(opts.base_branch, opts.target_branch)
    }

    #[cfg(test)]
    pub fn test_review_with_time_provider(
        opts: impl Into<TestReviewParams>,
        time_provider: &dyn TimeProvider,
    ) -> Self {
        let opts = opts.into();
        Self::new_with_time_provider(opts.base_branch, opts.target_branch, time_provider)
    }

    pub async fn save(&self, pool: &SqlitePool) -> Result<(), sqlx::Error> {
        sqlx::query(
            r#"
            INSERT INTO reviews (id, created_at, updated_at, base_branch, target_branch)
            VALUES (?1, ?2, ?3, ?4, ?5)
            "#,
        )
        .bind(&self.id)
        .bind(self.created_at.to_rfc3339())
        .bind(self.updated_at.to_rfc3339())
        .bind(&self.base_branch)
        .bind(&self.target_branch)
        .execute(pool)
        .await?;
        Ok(())
    }

    pub async fn list_all(pool: &SqlitePool) -> Result<Vec<Review>, sqlx::Error> {
        let reviews = sqlx::query_as::<_, Review>(
            r#"
            SELECT id, created_at, updated_at, base_branch, target_branch
            FROM reviews
            ORDER BY created_at DESC
            "#,
        )
        .fetch_all(pool)
        .await?;
        Ok(reviews)
    }

    pub async fn find_by_id(pool: &SqlitePool, id: &str) -> Result<Option<Review>, sqlx::Error> {
        let review = sqlx::query_as::<_, Review>(
            r#"
            SELECT id, created_at, updated_at, base_branch, target_branch
            FROM reviews
            WHERE id = ?1
            "#,
        )
        .bind(id)
        .fetch_optional(pool)
        .await?;
        Ok(review)
    }

    pub async fn delete(&self, pool: &SqlitePool) -> Result<(), sqlx::Error> {
        sqlx::query(
            r#"
            DELETE FROM reviews
            WHERE id = ?1
            "#,
        )
        .bind(&self.id)
        .execute(pool)
        .await?;
        Ok(())
    }
}

#[cfg(test)]
pub struct TestReviewParams {
    pub base_branch: String,
    pub target_branch: String,
}

#[cfg(test)]
impl TestReviewParams {
    pub fn new() -> Self {
        Self {
            base_branch: "default".to_string(),
            target_branch: "default".to_string(),
        }
    }

    pub fn base_branch(mut self, base_branch: &str) -> Self {
        self.base_branch = base_branch.to_string();
        self
    }

    pub fn target_branch(mut self, target_branch: &str) -> Self {
        self.target_branch = target_branch.to_string();
        self
    }
}

#[cfg(test)]
impl Default for TestReviewParams {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
impl From<()> for TestReviewParams {
    fn from(_: ()) -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_utils::fixed_time;
    use crate::time_provider::MockTimeProvider;
    use sqlx::SqlitePool;

    async fn create_test_pool() -> SqlitePool {
        let pool = SqlitePool::connect("sqlite::memory:").await.unwrap();
        sqlx::migrate!().run(&pool).await.unwrap();
        pool
    }

    #[test]
    fn test_review_new() {
        let base_branch = "main".to_string();
        let target_branch = "feature/test".to_string();
        let review = Review::new(base_branch.clone(), target_branch.clone());

        assert!(!review.id.is_empty());
        assert_eq!(review.created_at, review.updated_at);
        assert_eq!(review.base_branch, base_branch);
        assert_eq!(review.target_branch, target_branch);

        // ID should be a valid UUID
        assert!(uuid::Uuid::parse_str(&review.id).is_ok());
    }

    #[test]
    fn test_review_new_with_mock_time() {
        let fixed_time = fixed_time();
        let time_provider = MockTimeProvider::new(fixed_time);

        let review = Review::new_with_time_provider(
            "default".to_string(),
            "default".to_string(),
            &time_provider,
        );

        assert!(!review.id.is_empty());
        assert_eq!(review.created_at, fixed_time);
        assert_eq!(review.updated_at, fixed_time);
        assert_eq!(review.created_at, review.updated_at);
        assert_eq!(review.base_branch, "default".to_string());
        assert_eq!(review.target_branch, "default".to_string());

        // ID should be a valid UUID
        assert!(uuid::Uuid::parse_str(&review.id).is_ok());
    }

    #[test]
    fn test_review_new_time_provider() {
        let base_branch = "develop".to_string();
        let target_branch = "feature/branch-support".to_string();
        let fixed_time = fixed_time();
        let time_provider = MockTimeProvider::new(fixed_time);

        let review = Review::new_with_time_provider(
            base_branch.clone(),
            target_branch.clone(),
            &time_provider,
        );

        assert!(!review.id.is_empty());
        assert_eq!(review.created_at, fixed_time);
        assert_eq!(review.updated_at, fixed_time);
        assert_eq!(review.base_branch, base_branch);
        assert_eq!(review.target_branch, target_branch);

        // ID should be a valid UUID
        assert!(uuid::Uuid::parse_str(&review.id).is_ok());
    }

    #[test]
    fn test_title() {
        let review = Review::test_review(());
        assert_eq!(review.title(), "default -> default");
        let custom_review = Review::test_review(
            TestReviewParams::new()
                .base_branch("main")
                .target_branch("feature/test"),
        );
        assert_eq!(custom_review.title(), "main -> feature/test");
    }

    #[tokio::test]
    async fn test_review_save_and_list() {
        let pool = create_test_pool().await;
        let review = Review::test_review(());

        // Save the review
        review.save(&pool).await.unwrap();

        // List all reviews
        let reviews = Review::list_all(&pool).await.unwrap();

        assert_eq!(reviews.len(), 1);
        assert_eq!(reviews[0].id, review.id);
        assert_eq!(reviews[0].base_branch, "default".to_string());
        assert_eq!(reviews[0].target_branch, "default".to_string());
    }

    #[tokio::test]
    async fn test_review_save_and_list_with_branches() {
        let pool = create_test_pool().await;
        let base_branch = "main".to_string();
        let target_branch = "feature/test".to_string();
        let review = Review::test_review(
            TestReviewParams::new()
                .base_branch(&base_branch)
                .target_branch(&target_branch),
        );

        // Save the review
        review.save(&pool).await.unwrap();

        // List all reviews
        let reviews = Review::list_all(&pool).await.unwrap();

        assert_eq!(reviews.len(), 1);
        assert_eq!(reviews[0].id, review.id);
        assert_eq!(reviews[0].base_branch, base_branch);
        assert_eq!(reviews[0].target_branch, target_branch);

        // Find by ID
        let found_review = Review::find_by_id(&pool, &review.id)
            .await
            .unwrap()
            .unwrap();
        assert_eq!(found_review.base_branch, base_branch);
        assert_eq!(found_review.target_branch, target_branch);
    }

    #[tokio::test]
    async fn test_review_list_empty() {
        let pool = create_test_pool().await;

        let reviews = Review::list_all(&pool).await.unwrap();

        assert_eq!(reviews.len(), 0);
    }

    #[tokio::test]
    async fn test_review_list_ordered_by_created_at_desc() {
        let pool = create_test_pool().await;

        let time1 = fixed_time();
        let time2 = time1 + chrono::Duration::hours(1);

        let time_provider1 = MockTimeProvider::new(time1);
        let time_provider2 = MockTimeProvider::new(time2);

        let review1 = Review::test_review_with_time_provider(
            TestReviewParams::default().base_branch("First Review"),
            &time_provider1,
        );
        let review2 = Review::test_review_with_time_provider(
            TestReviewParams::default().base_branch("Second Review"),
            &time_provider2,
        );

        // Save in order
        review1.save(&pool).await.unwrap();
        review2.save(&pool).await.unwrap();

        let reviews = Review::list_all(&pool).await.unwrap();

        assert_eq!(reviews.len(), 2);
        // Should be ordered by created_at DESC, so newest first
        assert_eq!(reviews[0].base_branch, "Second Review");
        assert_eq!(reviews[1].base_branch, "First Review");
        assert!(reviews[0].created_at > reviews[1].created_at);
    }

    #[tokio::test]
    async fn test_review_save_duplicate_id_fails() {
        let pool = create_test_pool().await;

        let review1 = Review::test_review(());
        let mut review2 = Review::test_review(());

        // Make them have the same ID
        review2.id = review1.id.clone();

        // First save should succeed
        review1.save(&pool).await.unwrap();

        // Second save with same ID should fail
        assert!(review2.save(&pool).await.is_err());
    }

    #[tokio::test]
    async fn test_review_find_by_id() {
        let pool = create_test_pool().await;
        let review = Review::test_review(());

        // Save the review
        review.save(&pool).await.unwrap();

        // Find by ID
        let found_review = Review::find_by_id(&pool, &review.id).await.unwrap();
        assert!(found_review.is_some());
        let found_review = found_review.unwrap();
        assert_eq!(found_review.id, review.id);
        assert_eq!(found_review.created_at, review.created_at);
        assert_eq!(found_review.updated_at, review.updated_at);

        // Find by non-existent ID
        let not_found = Review::find_by_id(&pool, "non-existent-id").await.unwrap();
        assert!(not_found.is_none());
    }

    #[tokio::test]
    async fn test_review_delete() {
        let pool = create_test_pool().await;
        let review = Review::test_review(());

        // Save the review
        review.save(&pool).await.unwrap();

        // Verify it exists
        let reviews = Review::list_all(&pool).await.unwrap();
        assert_eq!(reviews.len(), 1);

        // Delete the review
        review.delete(&pool).await.unwrap();

        // Verify it's gone
        let reviews = Review::list_all(&pool).await.unwrap();
        assert_eq!(reviews.len(), 0);
    }

    #[test]
    fn test_review_eq_same_id() {
        let time1 = fixed_time();
        let time2 = time1 + chrono::Duration::hours(1);

        let time_provider1 = MockTimeProvider::new(time1);
        let time_provider2 = MockTimeProvider::new(time2);

        let review1 = Review::test_review_with_time_provider(
            TestReviewParams::default().base_branch("main"),
            &time_provider1,
        );
        let mut review2 = Review::test_review_with_time_provider(
            TestReviewParams::default().base_branch("dev"),
            &time_provider2,
        );

        // Make review2 have the same ID as review1
        review2.id = review1.id.clone();

        // Should be equal despite different base_branch and timestamps
        assert_eq!(review1, review2);
    }

    #[test]
    fn test_review_eq_different_id() {
        let fixed_time = fixed_time();
        let time_provider = MockTimeProvider::new(fixed_time);

        let review1 = Review::test_review_with_time_provider(
            TestReviewParams::default().base_branch("Same Title"),
            &time_provider,
        );
        let review2 = Review::test_review_with_time_provider(
            TestReviewParams::default().base_branch("Same Title"),
            &time_provider,
        );

        // Should not be equal despite same base_branch and timestamps because IDs are different
        assert_ne!(review1, review2);
    }

    #[test]
    fn test_review_eq_self() {
        let review = Review::test_review(());

        // Should be equal to itself
        assert_eq!(review, review);
    }

    #[test]
    fn test_review_eq_clone() {
        let review1 = Review::test_review(());
        let review2 = review1.clone();

        // Clone should be equal to original
        assert_eq!(review1, review2);
    }

    #[test]
    fn test_review_test_helper() {
        // Test with all defaults
        let review1 = Review::test_review(());
        assert_eq!(review1.base_branch, "default");
        assert_eq!(review1.target_branch, "default");

        // Test with custom base_branch only
        let review2 = Review::test_review(TestReviewParams::new().base_branch("Custom Title"));
        assert_eq!(review2.base_branch, "Custom Title");
        assert_eq!(review2.target_branch, "default");

        // Test with all custom values
        let review3 = Review::test_review(
            TestReviewParams::new()
                .base_branch("main")
                .target_branch("feature/test"),
        );
        assert_eq!(review3.base_branch, "main");
        assert_eq!(review3.target_branch, "feature/test");

        // Test using Default trait
        let review4 = Review::test_review(TestReviewParams::default());
        assert_eq!(review4.base_branch, "default");
        assert_eq!(review4.target_branch, "default");
    }

    #[test]
    fn test_review_eq_ignores_other_fields() {
        let time1 = fixed_time();
        let time2 = time1 + chrono::Duration::days(30);

        let time_provider1 = MockTimeProvider::new(time1);

        let review1 = Review::new_with_time_provider(
            "default".to_string(),
            "default".to_string(),
            &time_provider1,
        );
        let review2 = Review {
            id: review1.id.clone(),                        // Same ID
            created_at: time2,                             // Different created_at
            updated_at: time2,                             // Different updated_at
            base_branch: "different-base".to_string(),     // Different base_branch
            target_branch: "different-target".to_string(), // Different target_branch
        };

        // Should be equal because only ID matters for equality
        assert_eq!(review1, review2);
    }
}
