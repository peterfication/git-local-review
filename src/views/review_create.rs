use crate::{
    app::App,
    event::AppEvent,
    services::review_service::ReviewCreateData,
    views::{ViewHandler, ViewType},
};
use ratatui::{
    buffer::Buffer,
    crossterm::event::{KeyCode, KeyEvent},
    layout::{Constraint, Direction, Layout, Rect},
    style::{Color, Style},
    widgets::{Block, BorderType, Clear, Paragraph, Widget},
};

#[derive(Default)]
pub struct ReviewCreateView {
    pub title_input: String,
}

impl ViewHandler for ReviewCreateView {
    fn view_type(&self) -> ViewType {
        ViewType::ReviewCreate
    }

    fn handle_key_events(&mut self, app: &mut App, key_event: KeyEvent) -> color_eyre::Result<()> {
        match key_event.code {
            KeyCode::Esc => {
                self.title_input.clear();
                app.events.send(AppEvent::ReviewCreateClose);
            }
            KeyCode::Enter => {
                app.events
                    .send(AppEvent::ReviewCreateSubmit(ReviewCreateData {
                        title: self.title_input.clone(),
                    }));
                self.title_input.clear();
            }
            KeyCode::Char(char) => {
                self.title_input.push(char);
            }
            KeyCode::Backspace => {
                self.title_input.pop();
            }
            _ => {}
        }
        Ok(())
    }

    fn render(&self, _app: &App, area: Rect, buf: &mut Buffer) {
        let popup_area = centered_rect(60, 40, area);

        Clear.render(popup_area, buf);

        let block = Block::bordered()
            .title("Create New Review")
            .border_type(BorderType::Rounded)
            .style(Style::default().bg(Color::Black));

        let inner = block.inner(popup_area);
        block.render(popup_area, buf);

        let chunks = Layout::default()
            .direction(Direction::Vertical)
            .constraints([
                Constraint::Length(1),
                Constraint::Length(3),
                Constraint::Length(1),
            ])
            .split(inner);

        let title_label = Paragraph::new("Title:");
        title_label.render(chunks[0], buf);

        let title_input = Paragraph::new(self.title_input.as_str())
            .block(Block::bordered())
            .style(Style::default().fg(Color::Yellow));
        title_input.render(chunks[1], buf);

        let help = Paragraph::new("Press Enter to create, Esc to cancel")
            .style(Style::default().fg(Color::Gray));
        help.render(chunks[2], buf);
    }

    #[cfg(test)]
    fn debug_state(&self) -> String {
        format!("ReviewCreateView(title_input: \"{}\")", self.title_input)
    }

    #[cfg(test)]
    fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
        self
    }
}

fn centered_rect(percent_x: u16, percent_y: u16, r: Rect) -> Rect {
    let popup_layout = Layout::default()
        .direction(Direction::Vertical)
        .constraints([
            Constraint::Percentage((100 - percent_y) / 2),
            Constraint::Percentage(percent_y),
            Constraint::Percentage((100 - percent_y) / 2),
        ])
        .split(r);

    Layout::default()
        .direction(Direction::Horizontal)
        .constraints([
            Constraint::Percentage((100 - percent_x) / 2),
            Constraint::Percentage(percent_x),
            Constraint::Percentage((100 - percent_x) / 2),
        ])
        .split(popup_layout[1])[1]
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::database::Database;
    use crate::event::{AppEvent, Event};
    use crate::models::review::Review;
    use crate::services::ReviewsLoadingState;
    use crate::test_utils::render_app_to_terminal_backend;
    use insta::assert_snapshot;
    use sqlx::SqlitePool;

    use ratatui::crossterm::event::{KeyCode, KeyEvent, KeyEventKind, KeyEventState, KeyModifiers};

    async fn create_test_app() -> App {
        let pool = SqlitePool::connect("sqlite::memory:").await.unwrap();
        Review::create_table(&pool).await.unwrap();

        let database = Database::from_pool(pool);
        let reviews = vec![];

        App {
            running: true,
            events: crate::event::EventHandler::new_for_test(),
            database,
            reviews,
            reviews_loading_state: ReviewsLoadingState::Loaded,
            view_stack: vec![],
        }
    }

    #[test]
    fn test_review_create_view_default() {
        let view = ReviewCreateView::default();
        assert_eq!(view.title_input, "");
    }

    #[tokio::test]
    async fn test_review_create_view_handle_char_input() {
        let mut app = create_test_app().await;
        let mut view = ReviewCreateView::default();

        let key_event = KeyEvent {
            code: KeyCode::Char('a'),
            modifiers: KeyModifiers::empty(),
            kind: KeyEventKind::Press,
            state: KeyEventState::empty(),
        };

        view.handle_key_events(&mut app, key_event).unwrap();

        assert_eq!(view.title_input, "a");
    }

    #[tokio::test]
    async fn test_review_create_view_handle_multiple_chars() {
        let mut app = create_test_app().await;
        let mut view = ReviewCreateView::default();

        let chars = ['H', 'e', 'l', 'l', 'o'];
        for c in chars {
            let key_event = KeyEvent {
                code: KeyCode::Char(c),
                modifiers: KeyModifiers::empty(),
                kind: KeyEventKind::Press,
                state: KeyEventState::empty(),
            };
            view.handle_key_events(&mut app, key_event).unwrap();
        }

        assert_eq!(view.title_input, "Hello");
    }

    #[tokio::test]
    async fn test_review_create_view_handle_backspace() {
        let mut app = create_test_app().await;
        let mut view = ReviewCreateView {
            title_input: "Hello".to_string(),
        };

        let key_event = KeyEvent {
            code: KeyCode::Backspace,
            modifiers: KeyModifiers::empty(),
            kind: KeyEventKind::Press,
            state: KeyEventState::empty(),
        };

        view.handle_key_events(&mut app, key_event).unwrap();

        assert_eq!(view.title_input, "Hell");
    }

    #[tokio::test]
    async fn test_review_create_view_handle_backspace_empty() {
        let mut app = create_test_app().await;
        let mut view = ReviewCreateView::default();

        let key_event = KeyEvent {
            code: KeyCode::Backspace,
            modifiers: KeyModifiers::empty(),
            kind: KeyEventKind::Press,
            state: KeyEventState::empty(),
        };

        view.handle_key_events(&mut app, key_event).unwrap();

        assert_eq!(view.title_input, "");
    }

    #[tokio::test]
    async fn test_review_create_view_handle_esc() {
        let mut app = create_test_app().await;
        let mut view = ReviewCreateView {
            title_input: "Some input".to_string(),
        };
        assert!(!app.events.has_pending_events());

        let key_event = KeyEvent {
            code: KeyCode::Esc,
            modifiers: KeyModifiers::empty(),
            kind: KeyEventKind::Press,
            state: KeyEventState::empty(),
        };

        view.handle_key_events(&mut app, key_event).unwrap();

        // Title input should be cleared
        assert_eq!(view.title_input, "");

        // Verify that a ReviewCreateClose event was sent
        assert!(app.events.has_pending_events());
        let event = app.events.try_recv().unwrap();
        assert!(matches!(event, Event::App(AppEvent::ReviewCreateClose)));
    }

    #[tokio::test]
    async fn test_review_create_view_handle_enter() {
        let mut app = create_test_app().await;
        let mut view = ReviewCreateView {
            title_input: "Test Review".to_string(),
        };
        assert!(!app.events.has_pending_events());

        let key_event = KeyEvent {
            code: KeyCode::Enter,
            modifiers: KeyModifiers::empty(),
            kind: KeyEventKind::Press,
            state: KeyEventState::empty(),
        };

        view.handle_key_events(&mut app, key_event).unwrap();

        // Title input should be cleared after submit
        assert_eq!(view.title_input, "");

        // Verify that a ReviewCreateSubmit event was sent with the correct title
        assert!(app.events.has_pending_events());
        let event = app.events.try_recv().unwrap();
        if let Event::App(AppEvent::ReviewCreateSubmit(data)) = event {
            assert_eq!(data.title, "Test Review");
        } else {
            panic!("Expected ReviewCreateSubmit event");
        }
    }

    #[tokio::test]
    async fn test_review_create_view_handle_enter_empty() {
        let mut app = create_test_app().await;
        let mut view = ReviewCreateView::default();

        let key_event = KeyEvent {
            code: KeyCode::Enter,
            modifiers: KeyModifiers::empty(),
            kind: KeyEventKind::Press,
            state: KeyEventState::empty(),
        };

        view.handle_key_events(&mut app, key_event).unwrap();

        // Should still work with empty input
        assert_eq!(view.title_input, "");
    }

    #[tokio::test]
    async fn test_review_create_view_handle_unknown_key() {
        let mut app = create_test_app().await;
        let mut view = ReviewCreateView::default();
        let initial_input = "Test".to_string();
        view.title_input = initial_input.clone();

        let key_event = KeyEvent {
            code: KeyCode::F(1),
            modifiers: KeyModifiers::empty(),
            kind: KeyEventKind::Press,
            state: KeyEventState::empty(),
        };

        view.handle_key_events(&mut app, key_event).unwrap();

        // Unknown keys should not change input
        assert_eq!(view.title_input, initial_input);
    }

    #[tokio::test]
    async fn test_review_create_view_render_default() {
        let view = ReviewCreateView::default();
        let app = App {
            view_stack: vec![Box::new(view)],
            ..create_test_app().await
        };

        assert_snapshot!(render_app_to_terminal_backend(app))
    }

    #[tokio::test]
    async fn test_review_create_view_render_with_title() {
        let view = ReviewCreateView {
            title_input: "My New Review".to_string(),
        };
        let app = App {
            view_stack: vec![Box::new(view)],
            ..create_test_app().await
        };

        assert_snapshot!(render_app_to_terminal_backend(app))
    }
}
