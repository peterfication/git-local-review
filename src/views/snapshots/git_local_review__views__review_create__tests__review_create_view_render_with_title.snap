---
source: src/views/review_create.rs
expression: render_app_to_terminal_backend(app)
---
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                ╭Create New Review─────────────────────────────╮                "
"                │Title:                                        │                "
"                │┌────────────────────────────────────────────┐│                "
"                ││My New Review                               ││                "
"                │└────────────────────────────────────────────┘│                "
"                │Press Enter to create, Esc to cancel          │                "
"                │                                              │                "
"                ╰──────────────────────────────────────────────╯                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
